[Sema] TF-288: Handle immutable properties in `Differentiable` derive…

dan-zheng opened the pull request:
…d conformances.

Previously, `Differentiable` derived conformances was not enabled for structs
with `let` stored properties with initial values. (This is because immutable stored
properties can never be updated.)

Now, immutable stored properties are implicitly marked with `@noDerivative`
and are diagnosed with a fixit.

Resolves [TF-288](https://bugs.swift.org/browse/TF-288).

dan-zheng committed:
[Sema] TF-288: Handle immutable properties in `Differentiable` derived conformances.

Previously, `Differentiable` derived conformances was not enabled for structs
with `let` stored properties with initial values. (This is because immutable
stored properties can never be updated.)

Now, immutable stored properties are implicitly marked with `@noDerivative`
and are diagnosed with a warning+fixit.

Resolves TF-288.

rxwei commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -2446,6 +2446,12 @@ WARNING(differentiable_implicit_noderivative_fixit,none,
       (Identifier))
+WARNING(differentiable_constant_property_implicit_noderivative_fixit,none,
+      "'let' properties with an initial value do not have a derivative; add "
+      "'@noDerivative' to make it explicit",
```
"initial value" -> "default value"

From https://docs.swift.org/swift-book/LanguageGuide/Properties.html:
> You can provide a default value for a stored property as part of its definition, as described in Default Property Values. You can also set and modify the initial value for a stored property during initialization. This is true even for constant stored properties, as described in Assigning Constant Properties During Initialization.

dan-zheng replied:
Nice! Done in https://github.com/apple/swift/pull/22905/commits/eb580db59d15b7dbe98178f350d680a59846a95b.

rxwei replied:
Note in this paragraph the last sentence is slightly confusing -- it seems to suggest overriding a default `let` with an initializer is possible.

dan-zheng replied:
Yes, that is confusing.
I wonder if the semantics are changed by [SE-0242](https://forums.swift.org/t/se-0242-synthesize-default-values-for-the-memberwise-initializer/20618?u=dan-zheng).

rxwei commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -2446,6 +2446,12 @@ WARNING(differentiable_implicit_noderivative_fixit,none,
       "stored property %0 has no derivative because it does not conform to "
```
Can you rename this to `differentiable_nondiff_type_implicit_noderivative_fixit`?

dan-zheng replied:
Done in https://github.com/apple/swift/pull/22905/commits/eb580db59d15b7dbe98178f350d680a59846a95b.

rxwei commented on lib/Sema/DerivedConformanceDifferentiable.cpp:
```diff
@@ -851,19 +853,36 @@ static void checkAndDiagnoseImplicitNoDerivative(TypeChecker &TC,
+      continue;
+    }
+    TC.diagnose(
+          loc,
```
Slightly awkward indentation.

dan-zheng replied:
I hate it too. The diagnostic name is too long, this is what's suggested by `clang-format`.

dan-zheng committed:
Minor edits.

Address feedback from @rxwei.

dan-zheng committed:
Update diagnostic message.

Clarify that `let` stored properties with a default value can be changed
to `var` to allow derivatives.

rxwei commented:
@swift-ci please test tensorflow
